File: src/editor/gui/CardTagPanel.java
package editor.gui;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextField;
import javax.swing.SwingUtilities;

import editor.database.card.Card;
import editor.gui.generic.ScrollablePanel;
import editor.gui.generic.TristateCheckBox;
import editor.gui.generic.TristateCheckBox.State;
import editor.gui.settings.SettingsBuilder;
import editor.gui.settings.SettingsDialog;
import editor.util.MouseListenerFactory;
import editor.util.UnicodeSymbols;

/**
 * This class represents a panel that displays a list of card tags with check boxes
 * next to them.  A check mark means all cards it was given have a tag, a "mixed"
 * icon means that some of them do, and an empty box means none of them do.  Boxes can
 * be selected to modify tags.  Tags are not actually modified by the panel, but
 * it provides functions indicating what has been modified so that other objects
 * can.
 *
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class CardTagPanel extends ScrollablePanel
{
    /**
     * Maximum amount of rows to display in a scroll pane.
     */
    private static final int MAX_PREFERRED_ROWS = 10;

    /**
     * Show a dialog allowing editing of the tags of the given cards and adding
     * new tags.
     *
     * @param cards cards whose tags should be edited
     */
    public static void editTags(List<Card> cards, Component parent)
    {
        JPanel contentPanel = new JPanel(new BorderLayout());
        CardTagPanel cardTagPanel = new CardTagPanel(cards);
        contentPanel.add(new JScrollPane(cardTagPanel), BorderLayout.CENTER);
        JPanel lowerPanel = new JPanel();
        lowerPanel.setLayout(new BoxLayout(lowerPanel, BoxLayout.X_AXIS));
        JTextField newTagField = new JTextField();
        lowerPanel.add(newTagField);
        JButton newTagButton = new JButton("Add");

        ActionListener addListener = (e) -> {
            if (!newTagField.getText().isEmpty())
            {
                if (cardTagPanel.addTag(newTagField.getText()))
                {
                    newTagField.setText("");
                    cardTagPanel.revalidate();
                    cardTagPanel.repaint();
                    SwingUtilities.getWindowAncestor(cardTagPanel).pack();
                }
            }
        };
        newTagButton.addActionListener(addListener);
        newTagField.addActionListener(addListener);
        lowerPanel.add(newTagButton);
        contentPanel.add(lowerPanel, BorderLayout.SOUTH);
        if (JOptionPane.showConfirmDialog(parent, contentPanel, "Edit Card Tags", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION)
        {
            var tags = new HashMap<>(SettingsDialog.settings().inventory.tags.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, (e) -> new HashSet<>(e.getValue()))));

            for (var entry : cardTagPanel.getTagged().entrySet())
                tags.compute(entry.getKey().multiverseid().get(0), (k, v) -> {
                    if (v == null)
                        v = new HashSet<>();
                    v.addAll(entry.getValue());
                    return v;
                });
            for (var entry : cardTagPanel.getUntagged().entrySet())
                tags.compute(entry.getKey().multiverseid().get(0), (k, v) -> {
                    if (v != null)
                    {
                        v.removeAll(entry.getValue());
                        if (v.isEmpty())
                            v = null;
                    }
                    return v;
                });
            SettingsDialog.applySettings(new SettingsBuilder(SettingsDialog.settings()).inventoryTags(tags).build());
        }
    }

    /**
     * Cards whose tags are to be modified.
     */
    private Collection<Card> cards;
    /**
     * Preferred viewport height of this panel.
     */
    private int preferredViewportHeight;
    /**
     * Boxes corresponding to tags.
     */
    private List<TristateCheckBox> tagBoxes;

    /**
     * Create a new CardTagPanel for editing the tags of the given collection
     * of cards.
     *
     * @param coll collection containing cards whose tags should be edited
     */
    public CardTagPanel(Collection<Card> coll)
    {
        super(TRACK_WIDTH);
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        setBackground(Color.WHITE);
        tagBoxes = new ArrayList<>();
        cards = coll;
        preferredViewportHeight = 0;

        setTags(SettingsDialog.settings().inventory.tags().stream().sorted().collect(Collectors.toList()));

        var tags = SettingsDialog.settings().inventory.tags.entrySet().stream().collect(Collectors.toMap((e) -> MainFrame.inventory().get(e.getKey()), Map.Entry::getValue));
        for (TristateCheckBox tagBox : tagBoxes)
        {
            long matches = cards.stream().filter((c) -> tags.get(c) != null && tags.get(c).contains(tagBox.getText())).count();
            if (matches == 0)
                tagBox.setState(State.UNSELECTED);
            else if (matches < cards.size())
                tagBox.setState(State.INDETERMINATE);
            else
                tagBox.setState(State.SELECTED);
        }
    }

    /**
     * Add a new tag to the list
     *
     * @param tag new tag to add
     * @return true if the tag was added, and false otherwise.
     */
    public boolean addTag(String tag)
    {
        var tags = tagBoxes.stream().map(TristateCheckBox::getText).collect(Collectors.toSet());
        if (tags.add(tag))
        {
            setTags(tags.stream().sorted().collect(Collectors.toList()));
            return true;
        }
        else
            return false;
    }

    /**
     * {@inheritDoc}
     * The maximum size of this CardTagPanel is MAX_PREFERRED_ROWS rows.
     */
    @Override
    public Dimension getPreferredScrollableViewportSize()
    {
        if (tagBoxes.isEmpty())
            return getPreferredSize();
        else
        {
            Dimension size = getPreferredSize();
            size.height = preferredViewportHeight;
            return size;
        }
    }

    /**
     * Get the cards that were tagged.
     *
     * @return a map of cards onto the sets tags that were added to them.
     */
    public Map<Card, Set<String>> getTagged()
    {
        var tagged = new HashMap<Card, Set<String>>();
        for (Card card : cards)
            for (TristateCheckBox tagBox : tagBoxes)
                if (tagBox.getState() == State.SELECTED)
                    tagged.compute(card, (k, v) -> {
                        if (v == null)
                            v = new HashSet<>();
                        v.add(tagBox.getText());
                        return v;
                    });
        return tagged;
    }

    /**
     * Get the cards that had tags removed.
     *
     * @return a map of cards onto the sets of tags that were removed from them.
     */
    public Map<Card, Set<String>> getUntagged()
    {
        var untagged = new HashMap<Card, Set<String>>();
        for (Card card : cards)
            for (TristateCheckBox tagBox : tagBoxes)
                if (tagBox.getState() == State.UNSELECTED)
                    untagged.compute(card, (k, v) -> {
                        if (v == null)
                            v = new HashSet<>();
                        v.add(tagBox.getText());
                        return v;
                    });
        return untagged;
    }

    /**
     * Remove a tag from the list.
     *
     * @param tag tag to remove
     * @return true if the tag was removed, and false otherwise.
     */
    public boolean removeTag(String tag)
    {
        var tags = tagBoxes.stream().map(TristateCheckBox::getText).collect(Collectors.toSet());
        if (tags.remove(tag))
        {
            setTags(tags.stream().sorted().collect(Collectors.toList()));
            if (getParent() != null)
            {
                getParent().revalidate();
                getParent().repaint();
            }
            if (SwingUtilities.getWindowAncestor(this) != null)
                SwingUtilities.getWindowAncestor(this).pack();
            return true;
        }
        else
            return false;
    }

    /**
     * Refresh the tags displayed with the given list of tags.
     *
     * @param tags list of tags that should be displayed
     */
    private void setTags(List<String> tags)
    {
        tagBoxes = tagBoxes.stream().filter((t) -> tags.contains(t.getText())).collect(Collectors.toList());
        removeAll();
        preferredViewportHeight = 0;
        if (tags.isEmpty())
        {
            JLabel emptyLabel = new JLabel("<html><i>No tags have been created.</i></html>");
            add(emptyLabel);
        }
        else
        {
            for (String tag : tags)
            {
                JPanel tagPanel = new JPanel();
                tagPanel.setBackground(Color.WHITE);
                tagPanel.setLayout(new BorderLayout());

                final TristateCheckBox tagBox = tagBoxes.stream().filter((t) -> t.getText().equals(tag)).findFirst().orElse(new TristateCheckBox(tag));
                tagBox.setBackground(Color.WHITE);
                tagPanel.add(tagBox, BorderLayout.WEST);
                tagBoxes.add(tagBox);

                JLabel deleteButton = new JLabel(String.valueOf(UnicodeSymbols.MINUS) + " ");
                deleteButton.setForeground(Color.RED);
                deleteButton.addMouseListener(MouseListenerFactory.createPressListener((e) -> removeTag(tag)));
                tagPanel.add(deleteButton, BorderLayout.EAST);

                preferredViewportHeight = Math.min(preferredViewportHeight + tagPanel.getPreferredSize().height, tagPanel.getPreferredSize().height * MAX_PREFERRED_ROWS);
                add(tagPanel);
            }
        }
    }
}

File: src/main/scala/editor/gui/CardTagPanel.scala
package editor.gui

import editor.database.attributes.CardAttribute
import editor.database.card.Card
import editor.gui.generic.ScrollablePanel
import editor.gui.generic.TristateCheckBox
import editor.gui.generic.TristateCheckBoxState
import editor.unicode._
import editor.util.MouseListenerFactory

import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.Box
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTextField
import javax.swing.SwingUtilities
import javax.swing.UIManager

/**
 * Companion object including convenience methods for [[CardTagPanel]].
 * @author Alec Roelke
 */
object CardTagPanel {
  /**
   * Show a dialog including a [[CardTagPanel]] to allow the user to edit tags for a selection of cards. If the user confirms
   * their selection, tags are added, removed, and deleted accordingly.
   * 
   * @param cards cards whose tags should be edited
   * @param parent parent component of the dialog, used for positioning
   */
  def editTags(cards: Seq[Card], parent: Component) = {
    val contentPanel = JPanel(new BorderLayout());
    val cardTagPanel = CardTagPanel(cards);
    contentPanel.add(JScrollPane(cardTagPanel), BorderLayout.CENTER);
    val lowerPanel = Box(BoxLayout.X_AXIS);
    val newTagField = JTextField();
    lowerPanel.add(newTagField);
    val newTagButton = JButton("Add");

    val addListener: ActionListener = _ => {
      if (!newTagField.getText.isEmpty && cardTagPanel.addTag(newTagField.getText, true)) {
        newTagField.setText("");
        cardTagPanel.revalidate();
        cardTagPanel.repaint();
        SwingUtilities.getWindowAncestor(cardTagPanel).pack();
      }
    }
    newTagButton.addActionListener(addListener);
    newTagField.addActionListener(addListener);
    lowerPanel.add(newTagButton);
    contentPanel.add(lowerPanel, BorderLayout.SOUTH);
    if (JOptionPane.showConfirmDialog(parent, contentPanel, "Edit Card Tags", JOptionPane.OK_CANCEL_OPTION, JOptionPane.PLAIN_MESSAGE) == JOptionPane.OK_OPTION) {
      val tagged = cardTagPanel.tagged
      val untagged = cardTagPanel.untagged
      cards.foreach((c) => {
        tagged.foreach((tag) => CardAttribute.Tags.tags(c) += tag)
        untagged.foreach((tag) => CardAttribute.Tags.tags(c) -= tag)
      })
      CardAttribute.Tags.tags.foreach((x: (Card, collection.mutable.Set[String])) => x match { case (_, t) => t --= cardTagPanel.removed })
    }
  }
}

/**
 * Panel allowing the user to edit tags applied to a collection of cards, add new tags, and remove existing tags.
 * Tags are shown as check boxes, with those where some cards in the collection have them and some don't initially
 * shown as being in indeterminate state.  Any left in that state will remain unchanged when the tagging is finished.
 * Note that, after tags are applied, any tags that aren't applied to any cards will be entirely deleted.
 * 
 * @constructor create a new panel editing tags of a given set of cards
 * @param cards cards whose tags should be edited
 * 
 * @author Alec Roelke
 */
class CardTagPanel(cards: Iterable[Card]) extends ScrollablePanel(ScrollablePanel.TrackWidth) {
  private val MaxPreferredRows = 10

  setLayout(BoxLayout(this, BoxLayout.Y_AXIS))
  setBackground(Color.WHITE)

  private val tagBoxes = collection.mutable.ArrayBuffer[TristateCheckBox]()
  private val _removed = collection.mutable.HashSet[String]()
  private var preferredViewportHeight = 0

  setTags(CardAttribute.Tags.tags.flatMap{ case (_, s) => s }.toSet)

  tagBoxes.foreach((box) => {
    val matches = cards.count((c) => Option(CardAttribute.Tags.tags(c)).exists(_.contains(box.getText)))
    if (matches == 0)
      box.state = TristateCheckBoxState.Unselected
    else if (matches < cards.size)
      box.state = TristateCheckBoxState.Indeterminate
    else
      box.state = TristateCheckBoxState.Selected
  })

  private def setTags(tags: Set[String]): Unit = {
    tagBoxes --= tagBoxes.filter((box) => !tags.contains(box.getText))
    removeAll()
    preferredViewportHeight = 0
    if (tags.isEmpty)
      add(JLabel("<html><i>No tags have been created.</i></html>"))
    else {
      tags.toSeq.sorted.foreach((tag) => {
        val tagPanel = JPanel(BorderLayout())
        tagPanel.setBackground(UIManager.getColor("List.background"))

        val tagBox = tagBoxes.find(_.getText == tag).getOrElse(new TristateCheckBox(tag))
        tagBox.setBackground(tagPanel.getBackground)
        tagPanel.add(tagBox, BorderLayout.WEST)
        tagBoxes += tagBox

        val deleteButton = JLabel(s"$Minus ")
        deleteButton.setForeground(Color.RED)
        deleteButton.addMouseListener(MouseListenerFactory.createMouseListener(pressed = _ => removeTag(tag)))
        tagPanel.add(deleteButton, BorderLayout.EAST)

        preferredViewportHeight = Math.min(preferredViewportHeight + tagPanel.getPreferredSize.height, tagPanel.getPreferredSize.height*MaxPreferredRows)
        add(tagPanel)
      })
    }
  }

  /**
   * Add a new tag and optionally apply it to all the selected cards.
   * 
   * @param tag tag to add
   * @param selected whether or not it should be applied to all the cards
   * @return true if the tag was added, or false otherwise (because it was already there).
   * @note If the tag is added unselected and then the choice is confirmed, it won't be added to the tag pool.
   */
  def addTag(tag: String, selected: Boolean) = if (tagBoxes.exists(_.getText == tag)) false else {
    setTags((tagBoxes.map(_.getText) :+ tag).toSet)
    if (selected)
      tagBoxes.filter(_.getText == tag).foreach(_.setSelected(true))
    true
  }

  /**
   * Remove a tag entirely from the tag pool.  This differs from simply unchecking a tag box in that
   * it removes it from all cards and not just from the selected ones.
   * 
   * @param tag tag to remove
   * @return true if the tag was removed, and false otherwise (because it wasn't there to begin with)
   */
  def removeTag(tag: String) = if (!tagBoxes.exists(_.getText == tag)) false else {
    _removed += tag
    setTags((tagBoxes.collect{ case box if box.getText != tag => box.getText }).toSet)
    Option(getParent).foreach((p) => { p.validate(); p.repaint() })
    Option(SwingUtilities.getWindowAncestor(this)).foreach(_.pack())
    true
  }

  /** @return the set of tags that are checked (to be applied to the selected cards) */
  def tagged = tagBoxes.collect{ case box if box.state == TristateCheckBoxState.Selected => box.getText }.toSet

  /** @return the set of tags that are unchecked (to be removed from the selected cards) */
  def untagged = tagBoxes.collect{ case box if box.state == TristateCheckBoxState.Unselected => box.getText }.toSeq ++ _removed.toSet

  /** @return the set of tags that are to be removed from all cards, and hence deleted */
  def removed = _removed.toSet

  override def getPreferredScrollableViewportSize = if (tagBoxes.isEmpty) getPreferredSize else {
    val size = getPreferredSize
    size.height = preferredViewportHeight
    size
  }
}